crates/bgp-pkt/src/wire/serializer/nlri/bgp_ls.rs

```diff
@@ -0,0 +1,412 @@
+use netgauze_serde_macros::WritingError;
+use std::{io::Write, net::IpAddr};
+
+#[derive(WritingError, Eq, PartialEq, Clone, Debug)]
+pub enum BgpLsNlriWritingError {
+    StdIoError(#[from_std_io_error] String),
+    MultiTopologyIdWritingError(#[from] MultiTopologyIdWritingError),
+    RouteDistinguisherWritingError(#[from] RouteDistinguisherWritingError),
+}
+
+impl WritablePdu<BgpLsNlriWritingError> for BgpLsNlri {
+    const BASE_LENGTH: usize = 4; // nlri type u16 + total nlri length u16
+
+    fn len(&self) -> usize {
+        Self::BASE_LENGTH + self.path_id().map_or(0, |_| 4) + self.nlri().len()
+    }
+
+    fn write<T: Write>(&self, writer: &mut T) -> Result<(), BgpLsNlriWritingError>
+    where
+        Self: Sized,
+    {
+        if let Some(path_id) = self.path_id() {
+            writer.write_u32::<NetworkEndian>(path_id)?;
+        }
+        let nlri = self.nlri();
+
+        // do not count add_path length since it is before the tlv
+        let tlv_len = self.len() as u16 - self.path_id.map_or(0, |_| 4);
+        write_tlv_header(writer, nlri.raw_code(), tlv_len)?;
+
+        nlri.write(writer)?;
+
+        Ok(())
+    }
+}
+
+impl WritablePdu<BgpLsNlriWritingError> for BgpLsVpnNlri {
+    const BASE_LENGTH: usize = 12; // nlri type u16 + total nlri length u16 + rd 8 bytes
+
+    fn len(&self) -> usize {
+        Self::BASE_LENGTH + self.path_id().map_or(0, |_| 4) + self.value.len()
+    }
+
+    fn write<T: Write>(&self, writer: &mut T) -> Result<(), BgpLsNlriWritingError>
+    where
+        Self: Sized,
+    {
+        if let Some(path_id) = self.path_id() {
+            writer.write_u32::<NetworkEndian>(path_id)?;
+        }
+
+        // do not count add_path length since it is before the tlv
+        let tlv_len = self.len() as u16 - self.path_id.map_or(0, |_| 4);
+        write_tlv_header(writer, self.value.raw_code(), tlv_len)?;
+
+        self.rd.write(writer)?;
+        self.value.write(writer)?;
+
+        Ok(())
+    }
+}
+
+impl WritablePdu<BgpLsNlriWritingError> for BgpLsNlriValue {
+    const BASE_LENGTH: usize = 0;
+
+    fn len(&self) -> usize {
+        match self {
+            BgpLsNlriValue::Node(data) => data.len(),
+            BgpLsNlriValue::Link(data) => data.len(),
+            BgpLsNlriValue::Ipv4Prefix(data) => data.len(),
+            BgpLsNlriValue::Ipv6Prefix(data) => data.len(),
+            BgpLsNlriValue::Unknown { value, .. } => value.len(),
+        }
+    }
+
+    fn write<T: Write>(&self, writer: &mut T) -> Result<(), BgpLsNlriWritingError>
+    where
+        Self: Sized,
+    {
+        match self {
+            BgpLsNlriValue::Node(data) => data.write(writer),
+            BgpLsNlriValue::Link(data) => data.write(writer),
+            BgpLsNlriValue::Ipv4Prefix(data) => data.write(writer),
+            BgpLsNlriValue::Ipv6Prefix(data) => data.write(writer),
+            BgpLsNlriValue::Unknown { value, .. } => Ok(writer.write_all(value)?),
```

Comment:
Just checking, where the code is written?

Reply:
[crates/bgp-pkt/src/wire/serializer/nlri/bgp_ls.rs](https://github.com/NetGauze/NetGauze/pull/32/files#diff-275aea5683c16da22953fd40c0517fafbd8beb0318253069b7eafadc4c3c7cda)

line 58 and 83 for BgpLsNlri and BgpLsNlriVpn

Reply:
couldn't make a link to the specific lines haha sorry